### PR TITLE
V1.0.0

### PR DESCRIPTION
**Pull request by cat-haines**
- Fully namespaced constants
- Changed behaviour of enable/disable/size to bring inline with hardware.spiflash
- Significantly refactored _preverify and _postverify methods (where the most processing takes place)

The following table is the performance for 3x8K writes:

| Task | Original | First Refactor | Final Version | hardware.spiflash (reference |
| --- | --- | --- | --- | --- |
| no verification | 117ms | 92ms | 92ms | 110ms |
| SPIFLASH_PREVERIFY + SPIFLASH_POSTVERIFY | 4392ms | 1243ms | 481ms | 320ms |
| SPIFLASH_PREVERIFY | 2679ms | 666ms | 458ms | 181ms |
| SPIFLASH_POSTVERIFY | 2319ms | 665ms | 129ms | 183ms |

### CODE REVIEW

File: SpiFlash.class.nut [hunk 1]
```diff
@@ -1,68 +1,61 @@
...
+
+    // class members
     _spi = null;
     _cs_l = null;
     _blocks = null;
+    _enabled = null;
 
+    // aliased functions to speed things up
     _cs_l_w = null;
     _spi_w = null;
```
Comment (bbharris) [inline SpiFlash.class.nut]: Is this used anywhere?

File: SpiFlash.class.nut [hunk 2]
```diff
@@ -70,106 +63,101 @@ class SPIFlash {
...
+        // If we're already enabled, return
+        if (_enabled) return;
+
         _enabled = true;
 
         _cs_l_w(0);
-        _spi_w(RDP.tochar());
+        _spi_w(ELECTRICIMP_SPIFLASH_RDP);
         _cs_l_w(1);
     }
```
Comment (bbharris) [inline SpiFlash.class.nut]: I think you have to give the SPI some time to initialize after you enable before you can issue a command. Might want to throw a little blocking sleep in here to enforce that.

File: SpiFlash.class.nut (hunk 1 above)
Comment (bbharris) [inline SpiFlash.class.nut]: Doesn't seem to be initialized anywhere. How is this working at all?

**Comment by cat-haines** [comment]
@bbharris - I think you're looking at diff files, not the actual code files.. so bits probably appear like they're missing (because GitHub doesn't always show the entire file).

**Comment by cat-haines** [comment]
Any other comments on this PR, or are we happy to merge?

**Comment by blindman2k** [comment]
I would need to test it before I could say it was ready but I am happy with what we have so far. Any issues could come through as bug reports.

**Comment by blindman2k** [comment]
I just ran my battery of tests on this and it performs identically to the hardware.spiflash. My name is aron and I approve this message.

**Pull request merged**